### PR TITLE
[BCI-3989][core] - CR methods err when service unstarted

### PR DESCRIPTION
**Pull request by Farber98**
### Task Description:
CR methods should throw an error if service is unstarted

### This PR:
- Adds `StartChainReader` and `CloseChainReader` methods to the EVM `ChainReaderInterfaceTester`. 
- Check if the service is already started before calling `GetLatestValue`, `BatchGetLatestValues`, and `QueryKey`  
	- Didn't add the check to `Bind`, we can discuss later if it's needed there too.
	- Added an error case for each of the methods when calling without starting the service previously.
- Add `Start()` call to consumers of CR before using it.
- Modify EVM tests to include `StartChainReader` and `CloseChainReader` calls

### Ticket:
- [BCI-3989](https://smartcontract-it.atlassian.net/browse/BCI-3989)

### Depends on:
- https://github.com/smartcontractkit/chainlink-common/pull/705

[BCI-3989]: https://smartcontract-it.atlassian.net/browse/BCI-3989?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by Farber98** [comment]
Still need to
- Improve tests checking errors when calling the methods without the service started
- Find a workaround and prevent the breaking change + multiple repo PR

EDIT: 
- done
- couldn't

File: core/services/relay/evm/chain_reader.go [hunk 1]
```diff
@@ -40,10 +40,12 @@ type chainReader struct {
 	bindings
 	codec commontypes.RemoteCodec
 	commonservices.StateMachine
+	isStarted bool
```
Comment (EasterTheBunny) [inline core/services/relay/evm/chain_reader.go]: You should be able to access the started state in the `StateMachine` without the need to add a new property.

Comment (Farber98) [reply core/services/relay/evm/chain_reader.go]: Thanks for this! Looks cleaner. I was trying to replicate the way [txm components](https://github.com/smartcontractkit/chainlink/blob/v2.14.7-goreleaser-test17/common/txmgr/broadcaster.go#L150) do this

Comment (Farber98) [reply core/services/relay/evm/chain_reader.go]: changes addressed here: [`be1d2c8` (#14123)](https://github.com/smartcontractkit/chainlink/pull/14123/commits/be1d2c8f4b5d0cffe09eeca3e4f3129bfad3de84)

Comment (Farber98) [reply core/services/relay/evm/chain_reader.go]: I'll better use `StateMachine.Ready()`. Changes here: [`4ea601d` (#14123)](https://github.com/smartcontractkit/chainlink/pull/14123/commits/4ea601d1717f05d8944a16f2046711d2bda8a200)

Comment (Farber98) [reply core/services/relay/evm/chain_reader.go]: addressed here: [`19a0268` (#14123)](https://github.com/smartcontractkit/chainlink/pull/14123/commits/19a026846120854ec30bdcdacde7120db2d5637f)

File: core/services/registrysyncer/syncer.go [hunk 2]
```diff
@@ -276,6 +276,11 @@ func (s *registrySyncer) Sync(ctx context.Context, isInitialSync bool) error {
 			return err
 		}
 
+		err = reader.Start(ctx)
+		if err != nil {
+			return err
+		}
```
Comment (ilija42) [inline core/services/registrysyncer/syncer.go:283]: Shouldn't it be started in `func (s *registrySyncer) Start(ctx context.Context) error {`

File: core/services/relay/evm/chain_reader.go [hunk 3]
```diff
@@ -155,6 +160,10 @@ func (cr *chainReader) Start(ctx context.Context) error {
 
 // Close unregisters polling filters for bound contracts.
 func (cr *chainReader) Close() error {
+	if cr.StateMachine.State() == "Stopped" {
```
Comment (ilija42) [inline core/services/relay/evm/chain_reader.go]: Why is this necessary? Same for Start?

Comment (ilija42) [reply core/services/relay/evm/chain_reader.go]: Doesn't Start/StopOnce  cover this?

File: core/services/relay/evm/chain_reader.go [hunk 4]
```diff
@@ -141,6 +142,10 @@ func (cr *chainReader) Name() string { return cr.lggr.Name() }
 
 // Start registers polling filters if contracts are already bound.
 func (cr *chainReader) Start(ctx context.Context) error {
+	if cr.StateMachine.Ready() == nil {
```
Comment (ilija42) [inline core/services/relay/evm/chain_reader.go]: maybe upgrade to services.Engine as well

File: core/services/relay/evm/chain_reader.go (hunk 3 above)
Comment (jmank88) [reply core/services/relay/evm/chain_reader.go]: If this is being added to suppress a double-close, then we need to fix that root problem instead of patching it.

File: core/services/relay/evm/chain_reader.go [hunk 5]
```diff
@@ -176,6 +185,10 @@ func (cr *chainReader) HealthReport() map[string]error {
 }
 
 func (cr *chainReader) GetLatestValue(ctx context.Context, contractName, method string, confidenceLevel primitives.ConfidenceLevel, params, returnVal any) error {
+	if cr.StateMachine.Ready() != nil {
+		return errServiceNotStarted
+	}
```
Comment (jmank88) [inline core/services/relay/evm/chain_reader.go]: This presumes that the non-ready error is "not started" but that is not always the case. `StateMachine` has helpers that can be used instead. Something like:
```suggestion
	return cr.StateMachine.IfStarted(func() error {
		...
```

File: core/services/registrysyncer/syncer.go (hunk 2 above)
Comment (Farber98) [reply core/services/registrysyncer/syncer.go:283]: `func (s *registrySyncer) Start(ctx context.Context) error` calls `func (s *registrySyncer) syncLoop()` that calls `func (s *registrySyncer) Sync(ctx context.Context, isInitialSync bool) error` which initializes the CR. 

I added these lines to start the CR service right after this initialization

File: core/services/relay/evm/chain_reader.go (hunk 5 above)
Comment (Farber98) [reply core/services/relay/evm/chain_reader.go]: Cleaner! Changes here: [`2eb9b0d` (#14123)](https://github.com/smartcontractkit/chainlink/pull/14123/commits/2eb9b0db5364574716b8f9e7d5634c592fc3b149)

Comment (Farber98) [reply core/services/relay/evm/chain_reader.go]: What do you think about printing the `cr.StateMachine.State()` in the new changes when `cr.StateMachine.IfStarted` returns `!ok` ? 

Could it flip to `Started` between the exec and the logging becoming an inconsistent message? 

eg. [here](https://github.com/smartcontractkit/chainlink/blob/2eb9b0db5364574716b8f9e7d5634c592fc3b149/core/services/relay/evm/chain_reader.go#L196)

File: core/services/relay/evm/chain_reader.go (hunk 3 above)
Comment (Farber98) [reply core/services/relay/evm/chain_reader.go]: Fixed the root problem that was causing a second start attempt [`7448064` (#14123)](https://github.com/smartcontractkit/chainlink/pull/14123/commits/7448064f87e2cb8621e6375998840a337fe8731b)

File: core/services/relay/evm/chain_reader.go (hunk 4 above)
Comment (Farber98) [reply core/services/relay/evm/chain_reader.go]: @ilija42 What do you think about moving this into a separate ticket?

File: core/services/relay/evm/evmtesting/chain_reader_interface_tester.go [hunk 6]
```diff
@@ -231,11 +231,18 @@ func (it *EVMChainReaderInterfaceTester[T]) GetChainReader(t T) clcommontypes.Co
...
 	return cr
 }
 
+func (it *EVMChainReaderInterfaceTester[T]) Start(t T) {
+	require.NoError(t, it.cr.Start(it.Helper.Context(t)))
+}
+
+func (it *EVMChainReaderInterfaceTester[T]) Close(t T) {
+	require.NoError(t, it.cr.Close())
+}
```
Comment (silaslenihan) [inline core/services/relay/evm/evmtesting/chain_reader_interface_tester.go]: I'm kind of confused why this needs to be in the tester interface. We already have a `Setup` function which should start the `ChainReader` and close it upon cleanup. That way you wouldn't have to start and close it in each test since `Setup` is already being called.

Comment (Farber98) [reply core/services/relay/evm/evmtesting/chain_reader_interface_tester.go]: Thanks for this! Took a different approach, so we only expose `Close()`. 

The `Setup` function will start and stop the service in all the cases as you mention. We will use `Close()` only for cases where we want to verify that calling a method when the service is not in `Started` state returns an error

File: core/services/registrysyncer/syncer.go (hunk 2 above)
Comment (jmank88) [reply core/services/registrysyncer/syncer.go:283]: Why is it better to have the initialization here, rather than in `(*registrySyncer).Start`? Isn't that just contributing to this problem in the first place?

File: core/services/relay/evm/chain_reader.go (hunk 4 above)
Comment (ilija42) [reply core/services/relay/evm/chain_reader.go]: Sure

File: core/services/relay/evm/chain_reader.go (hunk 3 above)
Comment (ilija42) [reply core/services/relay/evm/chain_reader.go]: If the root problem is fixed do we still need this condition?

Comment (Farber98) [reply core/services/relay/evm/chain_reader.go]: It is not present now, the snippet is outdated

**Review by ilija42 — DISMISSED** [review]

**Review by silaslenihan — DISMISSED** [review]

File: core/services/registrysyncer/syncer.go (hunk 2 above)
Comment (Farber98) [reply core/services/registrysyncer/syncer.go:283]: What do you mean by `this problem`? 

There's a note in the `newReader` func that says why this is achieved this way: 
> // NOTE: this can't be called while initializing the syncer and needs to be called in the sync loop.
// This is because Bind() makes an onchain call to verify that the contract address exists, and if
// called during initialization, this results in a "no live nodes" error.

Comment (Farber98) [reply core/services/registrysyncer/syncer.go:283]: I added these lines as we are now enforcing `CR` service to be in `Started` state before calling the methods. This `syncer` was not starting the service previously and that's why we got some tests failing but now they are passing with this change

File: core/services/relay/evm/chain_reader.go (hunk 4 above)
Comment (Farber98) [reply core/services/relay/evm/chain_reader.go]: Created: https://smartcontract-it.atlassian.net/browse/BCI-4120

File: core/services/registrysyncer/syncer.go (hunk 2 above)
Comment (jmank88) [reply core/services/registrysyncer/syncer.go:283]: > What do you mean by this problem?

If all dependencies are `Started` before anything that uses them, then there is no need to be defensive about checking internally if `Start` has been called. But I see that there is a good reason to defer in this case.

File: go.mod [hunk 7]
```diff
@@ -349,10 +349,5 @@ require (
 	sigs.k8s.io/yaml v1.4.0 // indirect
 )
 
-replace (
-	// replicating the replace directive on cosmos SDK
-	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
-
-	// until merged upstream: https://github.com/mwitkow/grpc-proxy/pull/69
-	github.com/mwitkow/grpc-proxy => github.com/smartcontractkit/grpc-proxy v0.0.0-20230731113816-f1be6620749f
```
Comment (Farber98) [inline go.mod]: seems it's not needed anymore as it has been replaced by `github.com/smartcontractkit/grpc-proxy v0.0.0-20240830132753-a7e17fec5ab7`

File: go.md [hunk 8]
```diff
@@ -43,6 +43,8 @@ flowchart LR
   click chainlink-solana href "https://github.com/smartcontractkit/chainlink-solana"
   chainlink/v2 --> chainlink-starknet/relayer
   click chainlink-starknet/relayer href "https://github.com/smartcontractkit/chainlink-starknet"
+  chainlink/v2 --> grpc-proxy
```
Comment (Farber98) [inline go.md]: this file was making CI to fail after `github.com/smartcontractkit/grpc-proxy v0.0.0-20240830132753-a7e17fec5ab7` was added. 

`make generate` solved it by adding these lines

File: integration-tests/load/go.mod [hunk 9]
```diff
@@ -515,9 +515,6 @@ replace (
 	// replicating the replace directive on cosmos SDK
 	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
 
-	// until merged upstream: https://github.com/mwitkow/grpc-proxy/pull/69
-	github.com/mwitkow/grpc-proxy => github.com/smartcontractkit/grpc-proxy v0.0.0-20230731113816-f1be6620749f
-
```
Comment (ilija42) [inline integration-tests/load/go.mod:520]: Why did this get deleted?

File: integration-tests/load/go.mod [hunk 10]
```diff
@@ -515,9 +515,6 @@ replace (
 	// replicating the replace directive on cosmos SDK
 	github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.3-alpha.regen.1
 
-	// until merged upstream: https://github.com/mwitkow/grpc-proxy/pull/69
-	github.com/mwitkow/grpc-proxy => github.com/smartcontractkit/grpc-proxy v0.0.0-20230731113816-f1be6620749f
```
Comment (ilija42) [inline integration-tests/load/go.mod]: This should be merged soon https://github.com/smartcontractkit/chainlink/pull/14300/files

**Comment by Farber98** [comment]
Refactored the structure of `EVMInterfaceTester` for clarity and to make the [proposed solution](https://github.com/smartcontractkit/chainlink-common/pull/705#discussion_r1761094885) work:

- `Setup` receives the `started` flag to check if we should return CR started
- `Setup` divided into small functions to make it easier to understand the purpose/responsibility of each one.
- `Get` functions now only return the instance. `Setup` is handled separately
